sui_programmability/examples/crypto/sources/ec_ops.move

```diff
@@ -0,0 +1,443 @@
+        let points = vector::singleton(enc2.ciphertext);
+        vector::push_back(&mut points, enc1.ciphertext);
+        vector::push_back(&mut points, enc1.ephemeral);
+        vector::push_back(&mut points, *pk2);
+        let lhs = bls12381::g1_multi_scalar_multiplication(&scalars, &points);
+        if (!group_ops::equal(&lhs, &proof.a3)) {
+            return false
+        };
+
+        return true
+    }
+
+    #[test]
+    fun test_elgamal_ops() {
+        // We have two parties.
+        let (sk1, pk1) = insecure_elgamal_key_gen(2110);
+        let (_, pk2) = insecure_elgamal_key_gen(1021);
+        // A sender wishes to send an encrypted message to pk1.
+        let m = bls12381::g1_mul(&bls12381::scalar_from_u64(5555), &bls12381::g1_generator());
+        let enc1 = insecure_elgamal_encrypt(&pk1, 1234, &m);
+        // The first party decrypts the message.
+        let m1 = elgamal_decrypt(&sk1, &enc1);
+        assert_eq(m, m1);
+        // Now, the first party wishes to send an encrypted message to pk2.
+        let r2 = 4321;
+        let enc2 = insecure_elgamal_encrypt(&pk2, r2, &m);
+        // And to prove equality of the two encrypted messages.
+        let proof = insecure_equility_prove(&pk1, &pk2, &enc1, &enc2, &sk1,  8888, r2);
+        // Anyone can verify it.
+        assert!(equility_verify(&pk1, &pk2, &enc1, &enc2, &proof), 0);
+
+        // Proving with an invalid witness should result in a failed verification.
+        let bad_r2 = 1111;
+        let proof = insecure_equility_prove(&pk1, &pk2, &enc1, &enc2, &sk1, 8888, bad_r2);
+        assert!(!equility_verify(&pk1, &pk2, &enc1, &enc2, &proof), 0);
+    }
+
+    ///////////////////////////////////////
+    ////// tlock (an IBE decryption) //////
+
+    /// An encryption of 32 bytes message following https://eprint.iacr.org/2023/189.pdf.
+    struct IbeEncryption has store, drop, copy {
+        u: Element<bls12381::G2>,
+        v: vector<u8>,
+        w: vector<u8>,
+    }
+
+    public fun ibe_encryption_from_bytes(bytes: &vector<u8>): IbeEncryption {
+        assert!(vector::length(bytes) == 96 + 32 + 32, 0);
+        let buffer = vector::empty();
+        let i = 0;
+        while (i < 96) {
+            vector::push_back(&mut buffer, *vector::borrow(bytes, i));
+            i = i + 1;
+        };
+        let u = bls12381::g2_from_bytes(&buffer);
+
+        let v = vector::empty();
+        while (i < 96 + 32) {
+            vector::push_back(&mut v, *vector::borrow(bytes, i));
+            i = i + 1;
+        };
+
+        let w = vector::empty();
+        while (i < 96 + 32 + 32) {
+            vector::push_back(&mut w, *vector::borrow(bytes, i));
+            i = i + 1;
+        };
+
+        IbeEncryption { u, v, w }
+    }
+
+    // Encrypt a message 'm' for 'target'. Follows the algorithms of https://eprint.iacr.org/2023/189.pdf.
+    // Note that the algorithms in that paper use G2 for signatures, where the actual chain uses G1, thus
```

Comment:
let's use the name of the drand chain, so devs don't get confused (as we know drand has two mainnets).

Reply:
this is relevant to the function that sets up the inputs to encrypt, see `test_ibe_decrypt_drand` - there we have the concrete curl commands that were used to retrieve the public key and signature from the relevant chain